pkg/virt-handler/isolation/detector.go

```diff
@@ -143,20 +149,82 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 		if err != nil {
 			return err
 		}
-		rLimit := unix.Rlimit{
-			Max: uint64(memlockSize),
-			Cur: uint64(memlockSize),
-		}
-		err = prLimit(process.Pid(), unix.RLIMIT_MEMLOCK, &rLimit)
+		err = setProcessMemoryLockRLimit(process.Pid(), memlockSize)
 		if err != nil {
-			return fmt.Errorf("failed to set rlimit for memory lock: %v", err)
+			return fmt.Errorf("failed to set process %d memlock rlimit to %d: %v", process.Pid(), memlockSize, err)
 		}
 		// we assume a single process should match
 		break
 	}
 	return nil
 }
 
+const qemuProcessExecutable = "qemu-kvm"
+
+// AdjustQemuProcessMemoryLimits adjusts QEMU process MEMLOCK rlimits that runs inside
+// virt-launcher pod on the given VMI according to its spec.
+// Only VMI's with VFIO devices (e.g: SRIOV, GPU) require QEMU process MEMLOCK adjustment.
+func AdjustQemuProcessMemoryLimits(podIsoDetector PodIsolationDetector, vmi *v1.VirtualMachineInstance) error {
+	if !util.IsVFIOVMI(vmi) {
+		return nil
+	}
+
+	isolationResult, err := podIsoDetector.Detect(vmi)
+	if err != nil {
+		return err
+	}
+
+	processes, err := ps.Processes()
+	if err != nil {
+		return fmt.Errorf("failed to get all processes: %v", err)
+	}
+	qemuProcess := findIsolatedQemuProcess(processes, isolationResult.PPid())
+	if qemuProcess == nil {
+		return fmt.Errorf("QEMU process not found")
+	}
+	qemuProcessPid := qemuProcess.Pid()
+
+	memlockSize, err := getMemlockSize(vmi)
+	if err != nil {
+		return err
+	}
+
+	if err := setProcessMemoryLockRLimit(qemuProcessPid, memlockSize); err != nil {
+		return fmt.Errorf("failed to set process %d memlock rlimit to %d: %v", qemuProcessPid, memlockSize, err)
+	}
+	log.Log.V(5).Object(vmi).Infof("set process %+v memlock rlimits to: Cur: %d Max:%d",
+		qemuProcess, memlockSize, memlockSize)
+
+	return nil
+}
+
+// findIsolatedQemuProcess Returns the first occurrence of the QEMU process whose parent is PID"
+func findIsolatedQemuProcess(processes []ps.Process, pid int) ps.Process {
+	processes = childProcesses(processes, pid)
+	return lookupProcessByExecutable(processes, qemuProcessExecutable)
+}
```

Comment:
I don't see how looking up for the qemu process is different from looking up libvirtd in `AdjustResources`
I'm just curious why do we need this added complexity of this double loop of creating a slice of child processes then looping on it again to fin the executable?
This process looks very simple to me in `AdjustResources`

Reply:
> I don't see how looking up for the qemu process is different from looking up libvirtd in `AdjustResources`
> I'm just curious why do we need this added complexity of this double loop of creating a slice of child processes then looping on it again to fin the executable?
> This process looks very simple to me in `AdjustResources`

@vladikr  Although it look similar finding qemu process is different consider virt-launcher processes state:
```
pid: 2000 ppid: 100 cmd: virt-launcher
|
|_ pid: 3300 ppid: 2000 cmd: qemu-kvm 
|
|_ pid: 4400 ppid: 2000 cmd: virt-launcher   <-------------- isolation detector scope
    |
    |_pid 55000 ppid: 2000 cmd: libvirtd
 ```
- virt-launcher pod runs two virt-launcher processes, where `libvirtd` and `qemu` run's under different instance of virt-launcher process (== both has different parent process).
- Isolation detector, detects the nested virt-launcher process (parent of `libvirtd` process).
Finding `qemu` process requires finding the top-level virt-launcher process, get his child processes and finally look for `qemu-kvm` process.

Finding qemu process requires finding the top-level-virt-launcher pid (the detected virt-launcher process ppid), 
get his child processes and look for qemu-kvm process.

About the complexity,
In my opinion having `qemuProcesses` and `lookupProcessByExecutable` makes the code that looks for qemu process easier to understand (could easily reused for `AdjustResources`).
Doing so made it easier to add tests.

From what I understand and correct me if I am wrong, virt-launcher will run fixed number of processes
thus although having both loops, it shouldn't degrade performance in terms of big-O notation.

What you think?

Reply:
Thank you.